### DISCUSSION TITLE
Generation of Ultimate & IBNR based on BF Method

### DISCUSSION THREAD

**Discussion by muhammadshamsherkhan2002** [casact/chainladder-python#246] [Q&A]
I hope that you are fine and perfect.

I would like to ask you that what should I do for generation of Ultimate & IBNR based on the BornhuetterFerguson (BF) and how to put the bf rate.

Please help me in this regards.

**Comment by kennethshsu** [comment]
Were you having trouble following the tutorial [here](https://chainladder-python.readthedocs.io/en/stable/tutorials/deterministic-tutorial.html)?

**Reply by muhammadshamsherkhan2002** [reply]
Thanks kenneth for your guidance. 

By the tutorial, I got the logic for calculation of ultimate & ibnr based on BornhuetterFerguson (BF) model with having an apriori that taking only a single constant value for all origin periods. The code is here.

**code for BF Method**
- comauto = cl.load_sample("clrd").groupby("LOB").sum().loc["comauto"]
- bf_model = cl.BornhuetterFerguson(apriori=0.75)
- bf_model.fit(comauto["CumPaidLoss"], sample_weight=comauto["EarnedPremNet"].latest_diagonal)

In continuation of the above, please guide me further (may be I am wrong) that it is possible to generate bf model based on different apriori rates (origin period wise) while using chainladder-python library. Please go through the attached sample excel file (just for understanding purpose).

**There are two sheets in the excel file.**
1. calc. by single apriori rate
2. calc. by different apriori rate (period wise - where needed)  **--This is a requirement**

You are kindly requested to review the sheet and reply and guide me accordingly. Your reply will highly be appreciated.

[BF Method - Sample Working Sheet.xlsx](https://github.com/casact/chainladder-python/files/7895837/BF.Method.-.Sample.Working.Sheet.xlsx)

**Comment by jbogaardt** [comment]
It is possible. One just needs to exploit the fact that the underlying formulation of the BF apriori ultimate is `apriori * sample_weight`.
```python
comauto = cl.load_sample("clrd").groupby("LOB").sum().loc["comauto"]
# Ignore apriori rate
bf_model = cl.BornhuetterFerguson(apriori=1.0)
# Instead do everything in the sample_weight
sample_weight = (comauto["EarnedPremNet"].latest_diagonal*.75).trend(.05)
bf_model.fit(comauto["CumPaidLoss"], sample_weight=sample_weight)
```

**Reply by muhammadshamsherkhan2002** [reply]
Please further elaborate that from library, may I identify **BF Rate - as per origin wise** after providing apriori as .75 and trend as .05 which is being used for calculation of sample_weight in the above example?

**Comment by jbogaardt** [comment]
I don't understand the question. Trend is an example, but not necessary for all use-cases. The point that should be understood is that neither the `apriori` nor the `sample_weight` on its own needs to be defined purely in `BornhuetterFerguson`. What really matters is that `apriori * sample_weight` needs to be prior belief of the on the same basis as the ultimate your are estimating. Since the exposure, `sample_weight` , is a vector along the origin, it can vary along the origin however you wish.  This isn't unique to this library, it is how the Bornhuetter-Ferguson method works.
